java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -370,177 +363,226 @@ public final void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
-      currentStates = states;
-      scan(tree);
-      return currentStates;
+    private void evaluateConditionalOr(BinaryExpressionTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+      for (ExecutionState executionState : currentStates) {
+        for (ExecutionState conditionState : evaluateCondition(executionState, tree.leftOperand())) {
+          if (conditionState.getBooleanConstraint(conditionState.peek()) == SymbolicBooleanConstraint.TRUE) {
+            nextStates.addState(conditionState);
+          } else {
+            conditionState.pop();
+            nextStates.addAll(evaluateWithState(conditionState, tree.rightOperand()));
+          }
+        }
+      }
+      currentStates = nextStates;
     }
 
-    @Override
-    public void visitBlock(BlockTree tree) {
-      for (StatementTree statement : tree.body()) {
-        currentStates = evaluateStatement(currentStates, statement);
+    @CheckForNull
+    final Symbol.VariableSymbol extractVariableSymbol(ExpressionTree tree) {
+      Tree currentTree = tree;
+      if (isSuperOrThisMemberSelect(tree)) {
+        currentTree = ((MemberSelectExpressionTree) currentTree).identifier();
       }
+      if (currentTree.is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree identifierTree = (IdentifierTree) currentTree;
+        Symbol symbol = identifierTree.symbol();
+        if (symbol.isVariableSymbol()) {
+          return (Symbol.VariableSymbol) symbol;
+        }
+      }
+      return null;
     }
 
     @Override
     public void visitBreakStatement(BreakStatementTree tree) {
-      currentStates = PackedStatementStates.instantiateWithBreakStates(currentStates.states);
+      currentStates.breakStates.addAll(currentStates.states);
+      currentStates.states.clear();
+    }
+
+    @Override
+    public void visitCaseGroup(CaseGroupTree tree) {
+      scan(tree.body());
     }
 
     @Override
     public void visitContinueStatement(ContinueStatementTree tree) {
-      currentStates = PackedStatementStates.instantiate();
+      currentStates = new PackedStatementStates(currentStates);
     }
 
     @Override
     public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
       Set<Symbol.VariableSymbol> assignedSymbols = extractor.findAssignedVariables(tree);
       invalidateAssignedVariables(assignedSymbols);
-      currentStates = evaluateStatement(currentStates, tree.statement());
-      PackedStatementStates nextStates = PackedStatementStates.instantiate();
-      for (ExecutionState state : currentStates) {
-        if (evaluateExpression(state, tree.condition()) != SymbolicBooleanConstraint.TRUE) {
-          nextStates.addState(state);
+      scan(tree.statement());
+      scan(tree.condition());
+      for (ExecutionState executionState : currentStates) {
+        if (executionState.popLast() != SymbolicValue.BOOLEAN_TRUE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)